File: Sorting/SideMenuViewController.m
//
//  SideMenuViewController.m
//  Sorting
//
//  Created by Michael Woodruff on 29/10/2016.
//  Copyright © 2016 Mike Woodruff. All rights reserved.
//

#import "BubbleSortViewController.h"
#import "InsertionSortViewController.h"
#import "MergeSortViewController.h"
#import "QuickSortViewController.h"
#import "SideMenuViewController.h"

@interface SideMenuViewController ()

@property (nonatomic)  NSInteger selectedIndex;

@end

@implementation SideMenuViewController

- (void)viewDidLoad {
    
    [super viewDidLoad];
    self.selectedIndex = 0;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {

    if(indexPath.row != self.selectedIndex) {
        
        self.selectedIndex = indexPath.row;
    }
}

- (void)setSelectedIndex:(NSInteger)selectedIndex {
    
    _selectedIndex = selectedIndex;
    
    UIViewController* selectedViewController = nil;
    
    switch(selectedIndex) {
        case 0:
            selectedViewController = [[BubbleSortViewController alloc] init];
            break;
        case 1:
            selectedViewController = [[QuickSortViewController alloc] init];
            break;
        case 2:
            selectedViewController = [[MergeSortViewController alloc] init];
            break;
        case 3:
            selectedViewController = [[InsertionSortViewController alloc] init];
            break;
    }
    
    [self.delegate didSelectViewController:selectedViewController];
}

@end

File: Sorting/SideMenuViewController.swift
//
//  SideMenuViewController.swift
//  Sorting
//
//  Created by Michael Woodruff on 02/11/2016.
//  Copyright © 2016 Mike Woodruff. All rights reserved.
//

import Foundation
import UIKit

protocol SideMenuViewControllerDelegate: class {
    
    func didSelectViewController(viewController: UIViewController);
}

class SideMenuViewController: UITableViewController {
    
    weak var delegate: SideMenuViewControllerDelegate?;
    var selectedIndex: Int = 0 {
        didSet {
            didUpdateIndex();
        }
    }
    
    override func viewDidLoad() {
        super.viewDidLoad();
        selectedIndex = 0;
    }
    
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = super.tableView(tableView, cellForRowAt: indexPath);
        cell.textLabel?.font = .standardFont;
        return cell;
        
    }
    
    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        
        if(indexPath.row != self.selectedIndex) {
            
            selectedIndex = indexPath.row;
        }
    }
    
    internal func didUpdateIndex() {
        
        switch(selectedIndex) {
        case 0:
            
            delegate?.didSelectViewController(viewController: BubbleSortViewController());
            break;
        case 1:
            delegate?.didSelectViewController(viewController: QuickSortViewController());
            break;
        case 2:
            delegate?.didSelectViewController(viewController: MergeSortViewController());
            break;
        case 3:
            delegate?.didSelectViewController(viewController: InsertionSortViewController());
            break;
        default:
            delegate?.didSelectViewController(viewController: BubbleSortViewController());
            break;
        }
    }
}
